storm-core/src/clj/backtype/storm/daemon/nimbus.clj

```diff
@@ -346,53 +400,104 @@
+    (.createBlob blob-store conf-key (Utils/toCompressedJsonConf storm-conf) (SettableBlobMeta. BlobStoreAclHandler/DEFAULT) subject)
+    (if (instance? LocalFsBlobStore blob-store)
+      (.setup-blobstore! storm-cluster-state conf-key nimbus-host-port-info (get-metadata-version blob-store conf-key subject)))
+    (.createBlob blob-store code-key (Utils/serialize topology) (SettableBlobMeta. BlobStoreAclHandler/DEFAULT) subject)
+    (if (instance? LocalFsBlobStore blob-store)
+      (.setup-blobstore! storm-cluster-state code-key nimbus-host-port-info (get-metadata-version blob-store code-key subject)))))
+
+(defn- read-storm-topology [storm-id blob-store]
+  (Utils/deserialize
+    (.readBlob blob-store (master-stormcode-key storm-id) (get-subject)) StormTopology))
+
+(defn- get-blob-replication-count [blob-key nimbus]
+  (if (:blob-store nimbus)
+          (-> (:blob-store nimbus)
+            (.getBlobReplication  blob-key get-nimbus-subject)
+            (.get_replication))))
 
 (defn- wait-for-desired-code-replication [nimbus conf storm-id]
   (let [min-replication-count (conf TOPOLOGY-MIN-REPLICATION-COUNT)
         max-replication-wait-time (conf TOPOLOGY-MAX-REPLICATION-WAIT-TIME-SEC)
-        total-wait-time (atom 0)
-        current-replication-count (atom (if (:code-distributor nimbus) (.getReplicationCount (:code-distributor nimbus) storm-id) 0))]
-  (if (:code-distributor nimbus)
-    (while (and (> min-replication-count @current-replication-count)
-             (or (= -1 max-replication-wait-time)
-               (< @total-wait-time max-replication-wait-time)))
+        current-replication-count-jar (if (not (local-mode? conf)) (atom
+                                                                     (get-blob-replication-count (master-stormjar-key storm-id) nimbus))
+                                                                     (atom min-replication-count))
+        current-replication-count-code (atom (get-blob-replication-count (master-stormcode-key storm-id) nimbus))
+        current-replication-count-conf (atom (get-blob-replication-count (master-stormconf-key storm-id) nimbus))
+        total-wait-time (atom 0)]
+    (log-message "wait for desired replication" "count"
+      min-replication-count "wait-time" max-replication-wait-time
+      "code" @current-replication-count-code
+      "conf" @current-replication-count-conf
+      "jar" @current-replication-count-jar
+      "replication count" (get-blob-replication-count (master-stormconf-key storm-id) nimbus))
+    (if (:blob-store nimbus)
+      (while (and (> min-replication-count @current-replication-count-jar)
+               (> min-replication-count @current-replication-count-code)
+               (> min-replication-count @current-replication-count-conf)
+               (or (= -1 max-replication-wait-time)
+                 (< @total-wait-time max-replication-wait-time)))
```

Comment:
* Let's test for `(neg? max-replication-wait-time)`
* Shouldn't there be an or around the checks for jar, code, conf replication counts?  We want to wait until all of them have the right number of replications, not until any of them does:
```Clojure
(while
  (and 
    (or (> min-replication-count @current-replication-count-jar)
        (> min-replication-count @current-replication-count-code)
        (> min-replication-count @current-replication-count-conf))
    (or (neg? max-replication-wait-time)
        (< @total-wait-time max-replication-wait-time))))
```